### ISSUE TITLE
Better JACK support

### ISSUE THREAD

**Issue by tresf** [LMMS/lmms#1467]
This is a placeholder to consolidate JACK related issues and requests.  If a new JACK bug is reported on the tracker, close it out and mark it as a duplicate of this bug, #1467

---

**Enhancements:**
- [x] JACK support on Windows (#2136) and MacOS
- [ ] JACK transport support #523
- [x] JACK MIDI support #2038
- [ ] JACK inputs (for recording, eventually) #908
- [ ] JACK outputs #908
- [ ] JACK sends #524
- [ ] JACK freewheeling #908
- [ ] JACK samplerate should be default for exports #2211
- [x] Support Jack v1 and Jack v2 in our `AppImages` #3719
- [ ] Choosing how many outputs with JACK #5456

**Bugs:**
- [x] JACK output/rendering issues #1162
- [ ] JACK reconnection issues #1240

**Related:**
- [ ] ALSA to JACK bridge distortion #276

**Comment by Umcaruje** [comment]
This should be turned into a [task list](https://github.com/blog/1375%0A-task-lists-in-gfm-issues-pulls-comments) so its easier to mark fixed issues.

**Comment by tresf** [comment]
> This should be turned into a task list so its easier to mark fixed issues.

:)

**Comment by Reaper10** [comment]
#1977

**Comment by simonvanderveldt** [comment]
FYI Initial JACK midi support is in #2038

**Comment by unfa** [comment]
Does anyone know what is the status of JACK transport support?

It'd really kick ass to have this, even in a simplest form (like - no tempo changes supported).

**Comment by matt-bel** [comment]
**Enterprise** - Oggi alle 14:38
i checked out on this
seems to be possible
putting that when the play and stop button are played could be possible i think
**umcaruje** 💣 - Oggi alle 14:41
The problem comes when you play from the piano roll
Or the bb editor
Bitwig doesn't have this problem since they don't do this
I reckon transport should only happen when playing from the song editor
**Enterprise** - Oggi alle 14:42
having it just on the main play is it possible?
yes song editor play would be fine
**umcaruje** 💣 - Oggi alle 14:42
Everything is possible
You just have to code it well
Enterprise - Oggi alle 14:43
lol :smile:

**Comment by gi0e5b06** [comment]
@matt-bel https://github.com/gi0e5b06/lmms/commit/d8fb40cdb4943e4c0589281f8c3a4ad2f4a2dc0c
lmms as a client, basic transport.

**Comment by Umcaruje** [comment]
@gi0e5b06 here's how you make a Pull Request out of that commit:

First, you add an upstream remote (ignore if you did this)
```
$ git remote add upstream https://github.com/LMMS/lmms.git
```
You need to check out a new branch:
```
$ git checkout -b basicJackTransport
```
Next, we need to hard reset that branch to the current master:
```
$ git fetch upstream
$ git reset --hard upstream/master
```
Then you want to cherry pick your commit on the new branch:
```
$ git cherry-pick d8fb40cdb4943e4c0589281f8c3a4ad2f4a2dc0c
```
After that you can push your branch:
```
git push --set-upstream origin basicJackTransport
```
Then on the github UI on the website, when you go to https://github.com/LMMS/lmms, you'll see this:
![image](https://user-images.githubusercontent.com/6282045/32199064-4f9156c4-bdcb-11e7-99bb-4f5a6b85587c.png)
Click the green button to make a pull request.

And boom, your change gets reviewed and can safely get tested and merged.

**Comment by matt-bel** [comment]
@gi0e5b06 i've tryed to compile with it but that patch is not working, are you finishing it out or just abandoned?

**Comment by gi0e5b06** [comment]
The first patch was adding the `slave` mode (qjackctl controlling lmms).
The new patch (WiP) https://github.com/gi0e5b06/lmms/commit/45ccdc9e5b73c5ba9d38ea5890907d4e862837b9 adds the `client` mode (making the play/stop buttons in lmms working again) and introduces the ITransport interface.
I'm planning to add full `client` mode, however it is a bigger task because of the bad design.
I don't plan to add `master` mode soon.

**Comment by gi0e5b06** [comment]
@zonkmachine About the crash on exit (if you're using Jack), this is fixed by https://github.com/gi0e5b06/lmms/commit/b21a9cf3e03f49091cd064c6afde38708198f45f . Jack was still processing in the destructor.

@matt-bel About the sync: both Jack and LMMS being fully digital, they should stay synchronized (at the frame level). So that's something that can be fixed somehow. I put the transportQuery() call at the end of the processing. Let me know if it is better or worse.

**Comment by matt-bel** [comment]
@gi0e5b06 just tested it and tryed to sync with ardour metronome (only jack transport) with slave master and client option, still not perfectly synced, there's like a some millisecond delay, tested the other day (not this build but would be the same result) with jack transport + midi clock = https://www.youtube.com/watch?v=2PYt4hJUA9o, perfect because the midi delay control

**Comment by zonkmachine** [comment]
@gi0e5b06 Thanks! I've bumped the #developer on Discord.

**Comment by ycollet** [comment]
I vote for this one. Required to allow lmms to control xjadeo

**Comment by zicklag** [comment]
Sorry for the ping if commenting like this is discouraged, but I would love to have Jack transport support. I sync up MuseScore with Ardour to add digital instruments to my recordings, but LMMS has a lot more features from an effects and automation standpoint so it would give me a lot more powers if I could sync LMMS with Ardour.

**Comment by trebmuh** [comment]
@tresf : what about adding #5456 in https://github.com/LMMS/lmms/issues/1467#issue-52483742?

**Comment by Spekular** [comment]
@trebmuh done!

**Comment by lorenzosu** [comment]
Hi, I think by also reading other comments that Jack Transport is rather high on the list... I've commented on the specific issue which was closed (#523). Wondering if this is still being considered and what other users think.

**Comment by ycollet** [comment]
For me the issue with Jack is to be able to sync jack time between lmms and xjadeo to be able to compose music for the image.
Xjadeo is using this as sync source:
Sync sources:
- [JACK Transport](http://jackaudio.org/files/docs/html/transport-design.html)
- [Midi Timecode](http://en.wikipedia.org/wiki/MIDI_timecode) (MTC) via ALSA-MIDI, JACK-MIDI or portmidi
- [Linear (or Longitudinal) Timecode](http://en.wikipedia.org/wiki/Linear_timecode) (LTC) using JACK-audio

**Comment by tresf** [comment]
> Hi, I think by also reading other comments that Jack Transport is rather high on the list... I've commented on the specific issue which was closed (#523). Wondering if this is still being considered and what other users think.

We'd be happy to reopen any closed or consolidated bug reports if people want to start working on them.

**Comment by letsdig** [comment]
> For me the issue with Jack is to be able to sync jack time between lmms and xjadeo to be able to compose music for the image. Xjadeo is using this as sync source: Sync sources:
> 
> * [JACK Transport](http://jackaudio.org/files/docs/html/transport-design.html)
> * [Midi Timecode](http://en.wikipedia.org/wiki/MIDI_timecode) (MTC) via ALSA-MIDI, JACK-MIDI or portmidi
> * [Linear (or Longitudinal) Timecode](http://en.wikipedia.org/wiki/Linear_timecode) (LTC) using JACK-audio

This comment will not fit as issue answer and its maybe OT but as being requested from long time, this can be 'temporary(?)' solved in kinda couple ways: 

- Use carla as plugin (latest version), unblock extra features in carla preferences then go for the 'extra' tab and find the option to make the 'add jack app' tab appear. 
Now try to load xjadeo from the new tab 'Add jack application' , if everything works fine all should be synced with lmms.
Didn't tried it with xjadeo but i usually load jack apps into lmms->carla this way, using pulse for out in lmms, carla as plugin should virtualize a synced jack environment .

- Second way i solved this at this time is compiling/patchin from this fork: `https://github.com/firewall1110/lmms/tree/jackdexsync  `
Found that from this issue: `https://github.com/LMMS/lmms/pull/6066/commits/878a14ed64bf7d7d2c5c4dbafbea5c943cfc9892`
Then:
`https://github.com/LMMS/lmms/pull/6066`

Now i dont remember if i just compiled this or just patched a master branch version cause i have the patched archive saved in my local storage and also i've compiled my own appimage from it long ago, you should just try and just in case maybe dig a little bit.

The guy had the same kind of issue and solved this way for xjadeo...

Regards.

**Comment by Snowiiii** [comment]
Hey we updated jack to latest in https://github.com/LMMS/lmms/commit/3f5ac806e9c74048da6e2b0276ddc3581fdc575a does it may have improved the situation?

**Comment by letsdig** [comment]
why this would improve something never done?

**Comment by messmerd** [comment]
I've reopened all issues that were consolidated into this meta issue and marked as closed without actually being implemented. ([see this comment](https://github.com/LMMS/lmms/pull/7851#issuecomment-3621573077))
